[receiver/postgresql] Data race in integration test

https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5152433056/jobs/9278601236?pr=23025

```
2023/06/02 06:25:04 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 20.10.24+azure-1
  API Version: 1.41
  Operating System: Ubuntu 22.04.2 LTS
  Total Memory: 6943 MB
2023/06/02 06:25:04 🐳 Creating container for image docker.io/testcontainers/ryuk:0.4.0
2023/06/02 06:25:04 ✅ Container created: e03fdd721822
2023/06/02 06:25:04 🐳 Starting container: e03fdd7[218](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5152433056/jobs/9278601236?pr=23025#step:5:219)22
2023/06/02 06:25:04 🚧 Waiting for container id e03fdd721822 image: docker.io/testcontainers/ryuk:0.4.0
2023/06/02 06:25:04 ✅ Container started: e03fdd721822
2023/06/02 06:25:13 🐳 Creating container for image 
2023/06/02 06:25:13 ✅ Container created: 3a1361361c3e
2023/06/02 06:25:13 🐳 Starting container: 3a1361361c3e
2023/06/02 06:25:13 🚧 Waiting for container id 3a1361361c3e image: d47de94e-3639-4174-a20f-75ac3da7457d:dcbec011-b335-4fb6-a8c0-1e37ca7fc95d
2023/06/02 06:25:17 ✅ Container started: 3a1361361c3e
2023/06/02 06:25:20 🐳 Terminating container: 3a1361361c3e
2023/06/02 06:25:20 🚫 Container terminated: 3a1361361c3e
2023/06/02 06:25:25 🐳 Creating container for image 
2023/06/02 06:25:25 ✅ Container created: 43e03206e0ce
2023/06/02 06:25:25 🐳 Starting container: 43e03206e0ce
2023/06/02 06:25:25 🚧 Waiting for container id 43e03206e0ce image: b28c110f-915b-421f-a30b-73f74c15689c:2cec60e6-4e6f-4a67-b5b9-fbdd831452ce
2023/06/02 06:25:25 ✅ Container started: 43e03206e0ce
==================
WARNING: DATA RACE
Read at 0x00c000246480 by goroutine 324:
  go.opentelemetry.io/collector/receiver/scrapererror.(*ScrapeErrors).AddPartial()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.78.3-0.20230525165144-87dd85a6c034/scrapererror/scrapeerror.go:18 +0x30a
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).retrieveDatabaseSize()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:333 +0x2b3
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLClient).getDatabaseSize()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/client.go:189 +0x10c
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).retrieveDatabaseSize()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:330 +0x113
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).retrieveDBMetrics.func2()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:128 +0x10b

Previous write at 0x00c000246480 by goroutine 323:
  go.opentelemetry.io/collector/receiver/scrapererror.(*ScrapeErrors).AddPartial()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.78.3-0.20230525165144-87dd85a6c034/scrapererror/scrapeerror.go:18 +0x244
  go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.78.3-0.20230525165144-87dd85a6c034/scraperhelper/scrapercontroller.go:198 +0x34d
  go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.78.3-0.20230525165144-87dd85a6c034/scraperhelper/scrapercontroller.go:180 +0x184
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).collectWalAge()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:289 +0xa1
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).scrape()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:110 +0xafe
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).collectBGWriterStats()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:[220](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5152433056/jobs/9278601236?pr=23025#step:5:221) +0xa1
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).scrape()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:109 +0xac4
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).collectIndexes()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:197 +0xc4
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).scrape()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:105 +0x11db
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).collectTables()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:156 +0x2d7
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLClient).getBlocksReadByTable()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/client.go:293 +0xc4
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).collectTables()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:151 +0xc1
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).scrape()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:102 +0x1164
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).collectIndexes()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:197 +0xc4
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).scrape()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:105 +0x11db
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).collectTables()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:156 +0x2d7
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLClient).getBlocksReadByTable()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/client.go:293 +0xc4
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).collectTables()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:151 +0xc1
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).scrape()
      /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/postgresqlreceiver/scraper.go:102 +0x1164
  github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver.(*postgreSQLScraper).scrape-fm()
      <autogenerated>:1 +0x4d
  go.opentelemetry.io/collector/receiver/scraperhelper.ScrapeFunc.Scrape()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.78.3-0.20[230](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5152433056/jobs/9278601236?pr=23025#step:5:231)525165144-87dd85a6c034/scraperhelper/scraper.go:20 +0x6f
  go.opentelemetry.io/collector/receiver/scraperhelper.(*baseScraper).Scrape()
      <autogenerated>:1 +0x29
  go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.78.3-0.20230525165144-87dd85a6c034/scraperhelper/scrapercontroller.go:198 +0x34d
  go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1()
      /home/runner/go/pkg/mod/go.opentelemetry.io/collector/receiver@v0.78.3-0.20230525165144-87dd85a6c034/scraperhelper/scrapercontroller.go:180 +0x184
==================
2023/06/02 06:25:31 🐳 Terminating container: 43e03206e0ce
2023/06/02 06:25:31 🚫 Container terminated: 43e03206e0ce
2023/06/02 06:25:37 🐳 Creating container for image 
2023/06/02 06:25:37 ✅ Container created: ec333a3df04e
2023/06/02 06:25:37 🐳 Starting container: ec333a3df04e
2023/06/02 06:25:37 🚧 Waiting for container id ec333a3df04e image: 0c69a77c-a99b-430f-[873](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5152433056/jobs/9278601236?pr=23025#step:5:874)3-ada5680c5ed2:aedfb3e8-671b-47d2-b40d-f13c8415fd94
2023/06/02 06:25:41 ✅ Container started: ec333a3df04e
2023/06/02 06:25:44 🐳 Terminating container: ec333a3df04e
2023/06/02 06:25:44 🚫 Container terminated: ec333a3df04e
--- FAIL: TestIntegration (41.77s)
    --- FAIL: TestIntegration/multi_db (11.18s)
        testing.go:1319: race detected during execution of test
    testing.go:1319: race detected during execution of test
FAIL
```

--- a/.chloggen/postgres-int-race.yaml
+++ b/.chloggen/postgres-int-race.yaml
@@ -1,0 +1,20 @@
+# Use this changelog template to create an entry for release notes.
+# If your change doesn't affect end users, such as a test fix or a tooling change,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: postgresqlreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix race condition when capturing errors from multiple requests simultaneously
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [23026]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:

--- a/receiver/postgresqlreceiver/scraper.go
+++ b/receiver/postgresqlreceiver/scraper.go
@@ -25,6 +25,28 @@ type postgreSQLScraper struct {
 	clientFactory postgreSQLClientFactory
 	mb            *metadata.MetricsBuilder
 }
+type errsMux struct {
+	sync.RWMutex
+	errs scrapererror.ScrapeErrors
+}
+
+func (e *errsMux) add(err error) {
+	e.Lock()
+	defer e.Unlock()
+	e.errs.Add(err)
+}
+
+func (e *errsMux) addPartial(err error) {
+	e.Lock()
+	defer e.Unlock()
+	e.errs.AddPartial(1, err)
+}
+
+func (e *errsMux) combine() error {
+	e.Lock()
+	defer e.Unlock()
+	return e.errs.Combine()
+}
 
 type postgreSQLClientFactory interface {
 	getClient(c *Config, database string) (client, error)
@@ -83,7 +105,7 @@ func (p *postgreSQLScraper) scrape(ctx context.Context) (pmetric.Metrics, error)
 
 	now := pcommon.NewTimestampFromTime(time.Now())
 
-	var errs scrapererror.ScrapeErrors
+	var errs errsMux
 	r := &dbRetrieval{
 		activityMap: make(map[databaseName]int64),
 		dbSizeMap:   make(map[databaseName]int64),
@@ -94,7 +116,7 @@ func (p *postgreSQLScraper) scrape(ctx context.Context) (pmetric.Metrics, error)
 	for _, database := range databases {
 		dbClient, err := p.clientFactory.getClient(p.config, database)
 		if err != nil {
-			errs.Add(err)
+			errs.add(err)
 			p.logger.Error("Failed to initialize connection to postgres", zap.String("database", database), zap.Error(err))
 			continue
 		}
@@ -111,7 +133,7 @@ func (p *postgreSQLScraper) scrape(ctx context.Context) (pmetric.Metrics, error)
 	p.collectReplicationStats(ctx, now, listClient, &errs)
 	p.collectMaxConnections(ctx, now, listClient, &errs)
 
-	return p.mb.Emit(), errs.Combine()
+	return p.mb.Emit(), errs.combine()
 }
 
 func (p *postgreSQLScraper) retrieveDBMetrics(
@@ -119,7 +141,7 @@ func (p *postgreSQLScraper) retrieveDBMetrics(
 	listClient client,
 	databases []string,
 	r *dbRetrieval,
-	errs *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	wg := &sync.WaitGroup{}
 
@@ -147,15 +169,15 @@ func (p *postgreSQLScraper) recordDatabase(now pcommon.Timestamp, db string, r *
 	p.mb.EmitForResource(metadata.WithPostgresqlDatabaseName(db))
 }
 
-func (p *postgreSQLScraper) collectTables(ctx context.Context, now pcommon.Timestamp, dbClient client, db string, errs *scrapererror.ScrapeErrors) (numTables int64) {
+func (p *postgreSQLScraper) collectTables(ctx context.Context, now pcommon.Timestamp, dbClient client, db string, errs *errsMux) (numTables int64) {
 	blockReads, err := dbClient.getBlocksReadByTable(ctx, db)
 	if err != nil {
-		errs.AddPartial(1, err)
+		errs.addPartial(err)
 	}
 
 	tableMetrics, err := dbClient.getDatabaseTableMetrics(ctx, db)
 	if err != nil {
-		errs.AddPartial(1, err)
+		errs.addPartial(err)
 	}
 
 	for tableKey, tm := range tableMetrics {
@@ -192,11 +214,11 @@ func (p *postgreSQLScraper) collectIndexes(
 	now pcommon.Timestamp,
 	client client,
 	database string,
-	errs *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	idxStats, err := client.getIndexStats(ctx, database)
 	if err != nil {
-		errs.AddPartial(1, err)
+		errs.addPartial(err)
 		return
 	}
 
@@ -215,11 +237,11 @@ func (p *postgreSQLScraper) collectBGWriterStats(
 	ctx context.Context,
 	now pcommon.Timestamp,
 	client client,
-	errs *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	bgStats, err := client.getBGWriterStats(ctx)
 	if err != nil {
-		errs.AddPartial(1, err)
+		errs.addPartial(err)
 		return
 	}
 
@@ -243,11 +265,11 @@ func (p *postgreSQLScraper) collectMaxConnections(
 	ctx context.Context,
 	now pcommon.Timestamp,
 	client client,
-	errs *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	mc, err := client.getMaxConnections(ctx)
 	if err != nil {
-		errs.AddPartial(1, err)
+		errs.addPartial(err)
 		return
 	}
 	p.mb.RecordPostgresqlConnectionMaxDataPoint(now, mc)
@@ -257,11 +279,11 @@ func (p *postgreSQLScraper) collectReplicationStats(
 	ctx context.Context,
 	now pcommon.Timestamp,
 	client client,
-	errs *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	rss, err := client.getReplicationStats(ctx)
 	if err != nil {
-		errs.AddPartial(1, err)
+		errs.addPartial(err)
 		return
 	}
 	for _, rs := range rss {
@@ -284,7 +306,7 @@ func (p *postgreSQLScraper) collectWalAge(
 	ctx context.Context,
 	now pcommon.Timestamp,
 	client client,
-	errs *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	walAge, err := client.getLatestWalAgeSeconds(ctx)
 	if errors.Is(err, errNoLastArchive) {
@@ -292,7 +314,7 @@ func (p *postgreSQLScraper) collectWalAge(
 		return
 	}
 	if err != nil {
-		errs.AddPartial(1, fmt.Errorf("unable to determine latest WAL age: %w", err))
+		errs.addPartial(fmt.Errorf("unable to determine latest WAL age: %w", err))
 		return
 	}
 	p.mb.RecordPostgresqlWalAgeDataPoint(now, walAge)
@@ -304,13 +326,13 @@ func (p *postgreSQLScraper) retrieveDatabaseStats(
 	client client,
 	databases []string,
 	r *dbRetrieval,
-	errors *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	defer wg.Done()
 	dbStats, err := client.getDatabaseStats(ctx, databases)
 	if err != nil {
 		p.logger.Error("Errors encountered while fetching commits and rollbacks", zap.Error(err))
-		errors.AddPartial(1, err)
+		errs.addPartial(err)
 		return
 	}
 	r.Lock()
@@ -324,13 +346,13 @@ func (p *postgreSQLScraper) retrieveDatabaseSize(
 	client client,
 	databases []string,
 	r *dbRetrieval,
-	errors *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	defer wg.Done()
 	databaseSizeMetrics, err := client.getDatabaseSize(ctx, databases)
 	if err != nil {
 		p.logger.Error("Errors encountered while fetching database size", zap.Error(err))
-		errors.AddPartial(1, err)
+		errs.addPartial(err)
 		return
 	}
 	r.Lock()
@@ -344,12 +366,12 @@ func (p *postgreSQLScraper) retrieveBackends(
 	client client,
 	databases []string,
 	r *dbRetrieval,
-	errors *scrapererror.ScrapeErrors,
+	errs *errsMux,
 ) {
 	defer wg.Done()
 	activityByDB, err := client.getBackends(ctx, databases)
 	if err != nil {
-		errors.AddPartial(1, err)
+		errs.addPartial(err)
 		return
 	}
 	r.Lock()

